packages/core/src/common/message-rpc/rpc-message-encoder.ts

```diff
@@ -0,0 +1,477 @@
+    UNDEFINED = 4,
+    ObjectArray = 5,
+    RESPONSE_ERROR = 6,
+    ERROR = 7
+
+}
+
+/**
+ * A value encoder writes javascript values to a write buffer. Encoders will be asked
+ * in turn (ordered by their tag value, descending) whether they can encode a given value
+ * This means encoders with higher tag values have priority. Since the default encoders
+ * have tag values from 1-7, they can be easily overridden.
+ */
+export interface ValueEncoder {
+    /**
+     * Returns true if this encoder wants to encode this value.
+     * @param value the value to be encoded
+     */
+    is(value: any): boolean;
+    /**
+     * Write the given value to the buffer. Will only be called if {@link is(value)} returns true.
+     * @param buf The buffer to write to
+     * @param value The value to be written
+     * @param recursiveEncode A function that will use the encoders registered on the {@link MessageEncoder}
+     * to write a value to the underlying buffer. This is used mostly to write structures like an array
+     * without having to know how to encode the values in the array
+     */
+    write(buf: WriteBuffer, value: any, recursiveEncode?: (buf: WriteBuffer, value: any) => void): void;
+}
+
+/**
+ * Reads javascript values from a read buffer
+ */
+export interface ValueDecoder {
+    /**
+     * Reads a value from a read buffer. This method will be called for the decoder that is
+     * registered for the tag associated with the value encoder that encoded this value.
+     * @param buf The read buffer to read from
+     * @param recursiveDecode A function that will use the decoders registered on the {@link RpcMessageDecoder}
+     * to read values from the underlying read buffer. This is used mostly to decode structures like an array
+     * without having to know how to decode the values in the array.
+     */
+    read(buf: ReadBuffer, recursiveDecode: (buf: ReadBuffer) => unknown): unknown;
+}
+
+/**
+ * A `RpcMessageDecoder` parses a a binary message received via {@link ReadBuffer} into a {@link RpcMessage}
+ */
+export class RpcMessageDecoder {
+
+    protected decoders: Map<number, ValueDecoder> = new Map();
+    /**
+     * Declares the Uint8 type (i.e. the amount of bytes) necessary to store a decoder tag
+     * value in the buffer.
+     */
+    protected tagIntType: UintType;
+
+    constructor() {
+        this.registerDecoder(ObjectType.JSON, {
+            read: buf => JSON.parse(buf.readString())
+        });
+
+        this.registerDecoder(ObjectType.UNDEFINED, {
+            read: () => undefined
+        });
+
+        this.registerDecoder(ObjectType.ERROR, {
+            read: buf => {
+                const serializedError: SerializedError = JSON.parse(buf.readString());
+                const error = new Error(serializedError.message);
+                Object.assign(error, serializedError);
+                return error;
+            }
+        });
+
+        this.registerDecoder(ObjectType.RESPONSE_ERROR, {
```

Comment:
The previously used  `vscode-vs-jsonrpc` encoded all rpc errors as `ResponseError`s.  As a consequence the error handling for RPC calls in Theia services often expects that the received error is of `ResponseError` type ( or at least its often expected that the error has a `code` property). To not break the existing error handling we have to make sure that all errors that were previously received as `ResponseErrors` are also decoded to `ResponseErrors` with the new rpc-protocol.
We use the language server protocol type definition here because it is equivalent to the definition of `vscode-ws-json-rpc`.